apps/cow-fi/components/SearchBar/index.tsx

```diff
@@ -0,0 +1,284 @@
+import { useKeyboardNavigation } from '../../hooks/useKeyboardNavigation'
+
+interface SearchBarProps {
+  articles: Article[]
+}
+
+export const SearchBar: React.FC<SearchBarProps> = ({ articles }) => {
+  const [query, setQuery] = useState('')
+  const [filteredArticles, setFilteredArticles] = useState<Article[]>([])
+  const [isFocused, setIsFocused] = useState(false)
+  const [isPending, startTransition] = useTransition()
+  const [totalResults, setTotalResults] = useState(0)
+  const [currentPage, setCurrentPage] = useState(0)
+  const [hasMoreResults, setHasMoreResults] = useState(false)
+  const [isLoadingMore, setIsLoadingMore] = useState(false)
+  const [error, setError] = useState<string | null>(null)
+  const [showMinLengthMessage, setShowMinLengthMessage] = useState(false)
+  const inputRef = useRef<HTMLInputElement>(null)
+  const searchContainerRef = useRef<HTMLDivElement>(null)
+  const resultsRef = useRef<HTMLDivElement>(null)
+  const isMediumUp = useMediaQuery(Media.MediumAndUp(false))
+  const router = useRouter()
+
+  // Use custom debounce hook
+  const debouncedQuery = useDebounce(query, DEBOUNCE_DELAY)
+
+  // Handle selection for keyboard navigation
+  const handleSelect = (index: number) => {
+    if (index >= 0 && index < filteredArticles.length) {
+      const selectedArticle = filteredArticles[index]
+      router.push(`/learn/${selectedArticle.attributes?.slug}?ref=learn_search`)
+    }
+  }
+
+  // Handle escape key
+  const handleEscape = () => {
+    setIsFocused(false)
+    setQuery('')
+    setFilteredArticles([])
+    inputRef.current?.blur()
+  }
+
+  // Use custom keyboard navigation hook with improved focus management
+  const highlightedIndex = useKeyboardNavigation(
+    isFocused,
+    filteredArticles.length,
+    handleSelect,
+    handleEscape,
+    searchContainerRef,
+  )
+
+  // Server-side search using the searchArticlesAction
+  useEffect(() => {
+    const trimmedQuery = debouncedQuery.trim()
+
+    if (trimmedQuery.length > 0 && trimmedQuery.length < MIN_SEARCH_LENGTH) {
+      setShowMinLengthMessage(true)
+      setFilteredArticles([])
+      setTotalResults(0)
+      setHasMoreResults(false)
+      setError(null)
+      return
+    }
+
+    setShowMinLengthMessage(false)
+
+    if (trimmedQuery.length >= MIN_SEARCH_LENGTH) {
+      setCurrentPage(0) // Reset to first page on new search
+      setError(null) // Clear any previous errors
+
+      startTransition(async () => {
+        try {
+          const result = await searchArticlesAction(trimmedQuery, 0, PAGE_SIZE)
+          if (result.success && result.data) {
+            setFilteredArticles(result.data.data)
+            setTotalResults(result.data.meta.pagination.total)
+            setHasMoreResults(result.data.meta.pagination.pageCount > 1)
+          } else {
+            console.error('Search failed:', result.error)
+            setError('Unable to complete search. Please try again.')
+            setFilteredArticles([])
```

Comment:
```
setFilteredArticles([])
setTotalResults(0)
setHasMoreResults(false)
```

This code is duplicated 4 times, definitely deserves a separated function.

Reply:
Thanks for highlighting. Address this now.